base/sort.jl

```diff
@@ -426,7 +426,7 @@ function selectperm!{I<:Integer}(ix::AbstractVector{I}, v::AbstractVector,
                                  order::Ordering=Forward,
                                  initialized::Bool=false)
     if !initialized
-        @inbounds for i = 1:length(ix)
+        @inbounds for i in eachindex(ix)
```

Comment:
Shoot, clearly I need to expand the context when reviewing these (and perhaps not do it late at night). Yes, this falls under the "relying on the representation" clause.

Slightly radical thought, but I wonder if we should consider making all array indexes return their "context", so they can be converted into some kind of canonical form? For example, as an `immutable` with size information, so that linear<->cartesian transformations are possible when holding just the index. I don't think we want to go there yet, but it's something to keep in mind.

@KristofferC, I take your point. I don't want to make folks tackle too much, so in the future let's just run the benchmarks even on smaller PRs.

Also keep in mind that in many cases, this could become a performance win; in general, iteration with `eachindex` is no slower, and sometimes much much faster, than `for i = 1:n` iteration.